starters/express-typescript-typeorm-postgres/docker-compose.yml

```diff
@@ -0,0 +1,23 @@
+version: '3'
+services:
+  postgres:
+    image: 'postgres'
+    container_name: 'data_storage'
+    ports:
+      - '5432:5432'
+    volumes:
+      - ./pg_data:/var/lib/postgresql/data
+    env_file:
+      - ./postgres.db.env
+
+  redis_cache:
+    image: 'redis:alpine'
+    container_name: 'cache'
+    command: redis-server
+    ports:
+      - '6379:6379'
+    volumes:
+      - ./cache_data:/var/lib/redis
+      - ./cache_conf:/usr/local/etc/redis/redis.conf
```

Comment:
I had a lot of trouble with this.

When I ran `npm run dev` the first time, it seemed to work. 👍 

The seconds time, I got a bunch of errors:
```sh
[INFO] Example app listening on port 3333
[ERROR] [CACHE] An error occurred while connecting to Redis Error: connect ECONNREFUSED 127.0.0.1:6379
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 6379
}
Could not connect to the database, retrying 24 more time(s)
[INFO] [CACHE] Reconnecting to Redis instance
[ERROR] [CACHE] An error occurred while connecting to Redis Error: connect ECONNREFUSED 127.0.0.1:6379
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 6379
}
[INFO] [CACHE] Reconnecting to Redis instance
[ERROR] [CACHE] An error occurred while connecting to Redis Error: connect ECONNREFUSED 127.0.0.1:6379
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1157:16) {
  errno: -61,
  code: 'ECONNREFUSED',
  syscall: 'connect',
  address: '127.0.0.1',
  port: 6379
}
```

Then the next time, after looking in the docker dashboard, I got this weirdness:
```sh
[INFO] Example app listening on port 3333
[INFO] [CACHE] Successfully connected to Redis instance
[DEBUG] Connected to redis
Could not connect to the database, retrying 24 more time(s)
Could not connect to the database, retrying 23 more time(s)
Could not connect to the database, retrying 22 more time(s)
Could not connect to the database, retrying 21 more time(s)
Could not connect to the database, retrying 20 more time(s)
```

So, something is not right here. Is there maybe a race condition or something, the docker dashboard log said it couldn't find a config.

Reply:
That's strange. 🤔 I'm afraid we will have to look into it together, because when I do an `npm run infrastructure:clear` and then start it up again it works for me. The database does need polling for the first init, because the postgres docker image runs the init scripts when it is started up the first time and during that time it won't allow connections.

The Redis errors are valid if the Redis container didn't start up. Now that I thought it through I think if you don't have these docker images down, the first init will be slower. This might be solved with a script that pulls the images first.

I'll investigate some more. Do you think pulling the images could be part of a postinstall script? 🤔

Reply:
Personally I do not like postinstall scripts.

Reply:
So, I added a postinstall script, but I'm not convinced that I like it either.

I also tried running `dev` with a completely clean setup, with docker images removed and it started up for me. It had a little bit more boot up time, but eventually it got there. Now with a slow internet connection it also pose an issue, that is why I thought of the postinstall script.

But let's discuss it in the afternoon/your morning

Reply:
I also tried to "mitigate" the extreme amount of error logs coming from the API not being able to connect to the redis instance. Now it will log it once. I have a log for reconnecting, maybe that could be removed as well, or somehow throttled

Reply:
SO, whatever it was, running `npm run infrastructure:clear` seemed to fix it for me 🤷 

Maybe it was fine

Reply:
I'm glad it resolved. If you've checked out the branch before when only the database was set up and you created that docker image, that could have caused issues if the docker container wasn't cleared.

Nevertheless, I'll keep an eye out for similar errors. 👀